### DISCUSSION TITLE
v8.0.0-alpha.0

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/react-redux#1824] [General]
This is **the initial alpha preview release** for React-Redux v8.  

This alpha release reworks `useSelector` and `connect` for compatibility with React 18, rewrites the React-Redux codebase to TypeScript (obsoleting use of `@types/react-redux`), modernizes build output, and removes the deprecated `connectAdvanced` API and the `pure` option for `connect`.

```
npm i react-redux@next

yarn add react-redux@next
```

## Overview and Compatibility

**This alpha release is intended for initial compatibility testing with React 18 and TypeScript**.  For general discussion of this release, see the associated thread in the Discussions section.

Our public API is still the same ( `connect` and `useSelector/useDispatch`), all of our tests pass, and apps _should_ run okay.  It's very possible that the only migration needed is to bump the package version.

However, it's likely that there will be types breakage due to the TypeScript migration, and runtime bugs are also possible due to the amount of internal refactoring and changes.

React-Redux now requires the new experimental [`useSyncExternalStore` API in React 18](https://github.com/reactwg/react-18/discussions/86).  This release is using the "shim" package which backfills that API in earlier React versions, and currently lists `'react': '^16 || ^17 || 18'` as its acceptable peer dependencies, so in theory it could run with React 16 and 17.  However, we're planning to remove use of the shim and have a hard dependency on React 18 by the time version 8.0 goes final, so you are enouraged to try out this build with an "experimental" build of React 18 that contains `useSyncExternalStore`, such as version `0.0.0-experimental-7d38e4fd8-20210930`, and follow [the React 18 upgrade instructions](https://github.com/reactwg/react-18/discussions/6).

React 18 will be introducing new APIs related to SSR and hydration, and React-Redux will likely need further updates to support those. This release stubs out the `getServerSnapshot` argument to `useSyncExternalStore` - we'll tackle this in a future alpha release.

### TypeScript Migration

The React-Redux library source has always been written in plain JS, and the community maintained the TS typings separately as `@types/react-redux`.

We've (finally!) [migrated the React-Redux codebase to TypeScript](https://github.com/reduxjs/react-redux/issues/1737), using the existing typings as a starting point. We've tried to maintain the same external type signatures as much as possible, but there will most likely be some compile breakages from the changes, and we may have missed some bits along the way. Please file issues with any apparent TS-related problems so we can review them.

The TS migration was a great collaborative effort, with many community members contributing migrated files. Thank you to everyone who helped out!

As part of the process, we also updated the repo to use Yarn 2, copied the typetests files from DefinitelyTyped and expanded them, and improved our CI setup to test against multiple TS versions.

> **Note** When testing this out, you should remove the `@types/react-redux` package and ensure that any installed copies of `redux` are de-duped. You are also encouraged to update to the latest versions of Redux Toolkit (1.6.1+) or Redux (4.1.1), to ensure consistency between installed types and avoid problems from types mismatches.

### React 18 Compatibility

Per [the React 18 announcement overview](https://github.com/reactwg/react-18/discussions/4), React 18 will include new capabilities like automatic render batching and opt-in support for "concurrent rendering".  In order for external state libraries like Redux to take advantage of those, we need to use [the new `useSyncExternalStore` API](https://github.com/reactwg/react-18/discussions/86) to let React better coordinate renders caused by external updates.

We've reworked both `connect` and `useSelector` to call `useSyncExternalStore` internally. This process is purely internal refactoring, and should require no changes to your own code.

Early performance benchmarks show parity with React-Redux v7.2.5 for both `connect` and `useSelector`, so we do not anticipate any meaningful performance regressions.

### Modernized Build Output

We've always targeted ES5 syntax in our published build artifacts as the lowest common denominator. Even the "ES module" artifacts with `import/export` keywords still were compiled to ES5 syntax otherwise.

With IE11 now effectively dead and many sites no longer supporting it, we've updated our build tooling to target a more modern syntax equivalent to ES2017, which shrinks the bundle size slightly.

If you still need to support ES5-only environments, please compile your own dependencies as needed for your target environment.

### Removal of Legacy APIs

We announced in 2019 that [the legacy `connectAdvanced` API would be removed in the next major version](https://github.com/reduxjs/react-redux/issues/1236), as it was rarely used, added internal complexity, and was also basically irrelevant with the introduction of hooks.  As promised, we've removed that API.

We've also removed the `pure` option for `connect`, which forced components to re-render regardless of whether props/state had actually changed if it was set to `false`. This option was needed in some cases in the early days of the React ecosystem, when components sometimes relied on external mutable data sources that could change outside of rendering. Today, no one writes components that way, the option was barely used, and React 18's `useSyncExternalStore` strictly requires immutable updates.  So, we've removed the `pure` flag.

Given that both of these options were almost never used, this shouldn't meaningfully affect anyone.

## Changes

Due to the TS migration effort and number of contributors, this list covers just the major changes:

- [#1739: Integrate TypeScript port](https://github.com/reduxjs/react-redux/pull/1739)
- [#1808: Initial experimental React 18 compat prototyping](https://github.com/reduxjs/react-redux/pull/1808)
- [#1817: Fix compatibility with React 18 strict effects](https://github.com/reduxjs/react-redux/pull/1817)

https://github.com/reduxjs/react-redux/compare/v7.2.4...v8.0.0-alpha.0

<hr /><em>This discussion was created from the release <a href='https://github.com/reduxjs/react-redux/releases/tag/v8.0.0-alpha.0'>v8.0.0-alpha.0</a>.</em>

**Comment by einarq** [comment]
Which version of react and react-dom should I install to try this out?

I've installed react@next and react-dom@next, which gave me these:
    "react": "18.0.0-alpha-5fa4d79b0-20211008",
    "react-dom": "18.0.0-alpha-5fa4d79b0-20211008",

I still get the following error when starting the app:
Warning: You are using an outdated, pre-release alpha of React 18 that does not support useSyncExternalStore. The use-sync-external-store shim will not work correctly. Upgrade to a newer pre-release.
    at ConnectFunction (webpack-internal:///./node_modules/react-redux/es/components/connect.js:246:114)
    at Provider (webpack-internal:///./node_modules/react-redux/es/components/Provider.js:16:3)

I have these redux versions:
    "react-redux": "8.0.0-alpha.0",
    "redux": "4.1.1",

**Reply by markerikson** [reply]
Hmm. That... _ought_ to work as far as I know :)  Can you put up the repo?

**Comment by einarq** [comment]
That's difficult unfortunately, large enterprise app :)

I see from the code that this warnings gets thrown if React.startTransition
is defined. And I also see that React.startTransition is defined in my
version of React:
/./node_modules/react/cjs/react.development.js,with ReactVersion set
to '18.0.0-5fa4d79b0-20211008'

Do you have a different React version number I could try?

**Reply by markerikson** [reply]
Oh, you know what the issue probably is? React has both "alpha" and "experimental" builds, and `useSyncExternalStore` only exists in the "experimental" builds. You need one of those.  You can use the `experimental` tag, and as of right now the latest experimental version is `0.0.0-experimental-5fa4d79b0-20211008`.

**Comment by einarq** [comment]
Yup, that did the trick. Most things seem to just work, and some things
feel snappier (might be imagining things tho...).
There does seem to be some subtle changes in behavior tho, so I will have
to do some digging before we can eventually upgrade.

Anyway, very cool that there now is a viable upgrade path for React Redux,
thanks :)

**Reply by markerikson** [reply]
Cool. Can you add any further details on what "subtle changes in behavior" you're seeing?

**Comment by einarq** [comment]
Seem to be some smaller timing issues related to rendering and store updates. Keep in mind that this is a fairly large app, with sagas and a fair amount of slightly complicated rendering logic. Haven’t started really looking into making it work properly yet, just trying it out a bit. 

Is the expectation that things will work exactly as before? Using ReactDOM.createRoot btw, and the app is entirely StrictMode compatible.

**Reply by markerikson** [reply]
> Is the expectation that things will work exactly as before?

Well, that's sort of the point of all these "alphas" and "experimentals" :)  We _think_ things should generally work, but a lot of stuff has changed in both React and React-Redux.  So, it's entirely possible things may behave a bit differently.

For one thing, React now _always_ batches updates, which would be _a_ potential change with regards to render timing.  

I'd be very interested in any examples you can come up with where things behave differently between React 17 + React-Redux v7, and React 18 + React-Redux v8.
